### PR TITLE
Translations

### PR DESCRIPTION
**Pull request by toddmedema**

### CODE REVIEW

File: app/actions/Cards.tsx [hunk 1]
```diff
@@ -27,6 +27,17 @@ export function downloadCards(): ((dispatch: Redux.Dispatch<any>)=>void) {
...
         let cards: CardType[] = [];
         const sheets = tabletop.sheets();
+
+        let translations = null;
+        if (sheets.Translations) {
+          translations = sheets.Translations.elements.reduce((acculumator: TranslationsType, translation: {Language: string, Translated: string | boolean}) => {
+            acculumator[translation.Language] = translation.Translated;
+            return acculumator;
+          }, {});
+          delete sheets.Translations;
```
Comment (smartin015) [inline app/actions/Cards.tsx]: Why is this removed? For UI purposes? Suggest adding a comment.

File: app/helpers.tsx [hunk 2]
```diff
@@ -43,6 +44,33 @@ export function horizontalCounter(count: number): JSX.Element {
   return <span>{output}</span>;
 }
 
+// Returns the translated version of the string if available, otherwise the supplied English
+export function translate(english: string, translations: TranslationsType): string {
+  if (translations) {
+    if (translations[english]) {
+      return translations[english].toString();
+    } else {
+      console.log('Failed to locate translation for string: ' + english);
```
Comment (smartin015) [inline app/helpers.tsx]: `console.error`?

File: app/reducers/StateTypes.tsx [hunk 3]
```diff
@@ -3,9 +3,15 @@ export interface CardType {
   [key: string]: any;
 }
 
+export interface TranslationsType {
+  // Mapping of English string -> translation, or setting -> boolean
+  [key: string]: string | boolean;
```
Comment (smartin015) [inline app/reducers/StateTypes.tsx]: This would be a bit more rigorous if we explicitly listed boolean attributes (and assigned them when we parse the spreadsheet) instead of interleaving them here.

File: app/helpers.tsx [hunk 4]
```diff
@@ -43,6 +44,33 @@ export function horizontalCounter(count: number): JSX.Element {
   return <span>{output}</span>;
 }
 
+// Returns the translated version of the string if available, otherwise the supplied English
+export function translate(english: string, translations: TranslationsType): string {
+  if (translations) {
+    if (translations[english]) {
```
Comment (smartin015) [inline app/helpers.tsx]: Suggest lowercasing `translations` key strings at load time, and testing against `english.toLowerCase()` here. Otherwise a slight case mistake in a spreadsheet will break translation.

File: app/actions/Cards.tsx [hunk 5]
```diff
@@ -96,7 +116,7 @@ export function filterAndFormatCards(cards: CardType[], filters: FiltersState):
 
 
 const iconRegex = /#\w*/mg;
-const boldColonedRegex = /[^#]*: /g;
+const boldColonedRegex = /[^#:.\n]*?:/g;
```
Comment (smartin015) [inline app/actions/Cards.tsx]: What's the deal with this change? Was this not working properly before, or did it change due to the new translations? Do we need to strip newlines from inbound translated code?

Comment (toddmedema) [reply app/actions/Cards.tsx]: This was actually a separate bugfix, this was not working before on card contents that contain two :'s (specifically capsule of fate)

File: app/actions/Cards.tsx (hunk 1 above)
Comment (toddmedema) [reply app/actions/Cards.tsx]: Done

File: app/helpers.tsx (hunk 2 above)
Comment (toddmedema) [reply app/helpers.tsx]: Done

File: app/reducers/StateTypes.tsx (hunk 3 above)
Comment (toddmedema) [reply app/reducers/StateTypes.tsx]: Done

File: app/helpers.tsx (hunk 4 above)
Comment (toddmedema) [reply app/helpers.tsx]: Done
